pkg/controllers/nodeclaim/capacityreservation/controller.go

```diff
@@ -0,0 +1,151 @@
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	cpNodeClaims, err := c.cp.List(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+	}
+	cpNodeClaimIndex := lo.SliceToMap(cpNodeClaims, func(nc *karpv1.NodeClaim) (string, *karpv1.NodeClaim) {
+		return nc.Status.ProviderID, nc
+	})
+
+	ncs := &karpv1.NodeClaimList{}
+	if err := c.kubeClient.List(ctx, ncs); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodeclaims, %w", err)
+	}
+	updatedNodeClaims := sets.New[string]()
+	var errs []error
+	for i := range ncs.Items {
+		cpNC, ok := cpNodeClaimIndex[ncs.Items[i].Status.ProviderID]
+		if !ok {
+			continue
+		}
+		updated, err := c.syncCapacityType(ctx, cpNC.Labels[karpv1.CapacityTypeLabelKey], &ncs.Items[i])
+		if err != nil {
+			errs = append(errs, err)
+		}
+		if updated {
+			updatedNodeClaims.Insert(ncs.Items[i].Name)
+		}
+	}
+	log.FromContext(ctx).WithValues("NodeClaims", lo.Map(updatedNodeClaims.UnsortedList(), func(name string, _ int) klog.ObjectRef {
+		return klog.KRef("", name)
+	})).V(1).Info("updated capacity type for nodeclaims")
+
+	if len(errs) != 0 {
+		if lo.EveryBy(errs, func(err error) bool { return errors.IsConflict(err) }) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, multierr.Combine(errs...)
+	}
+	return reconcile.Result{RequeueAfter: time.Minute}, nil
+}
+
+// syncCapacityType will update the capacity type for the given NodeClaim. This accounts for the fact that capacity
+// reservations will expire, demoting NodeClaims with capacity type "reserved" to "on-demand".
+func (c *Controller) syncCapacityType(ctx context.Context, capacityType string, nc *karpv1.NodeClaim) (bool, error) {
+	// We won't be able to sync deleting NodeClaims, and there's no real need to either as they're already draining.
+	if !nc.DeletionTimestamp.IsZero() {
+		return false, nil
+	}
+
+	// For now we only account for the case where a reserved NodeClaim becomes an on-demand NodeClaim. This does not
+	// account for on-demand NodeClaims being promoted to reserved since that is not natively supported by Karpenter.
+	if capacityType != karpv1.CapacityTypeOnDemand {
+		return false, nil
+	}
+	if nc.Labels[karpv1.CapacityTypeLabelKey] == karpv1.CapacityTypeReserved {
+		stored := nc.DeepCopy()
+		nc.Labels[karpv1.CapacityTypeLabelKey] = karpv1.CapacityTypeOnDemand
+		delete(nc.Labels, cloudprovider.ReservationIDLabel)
+		if err := c.kubeClient.Patch(ctx, nc, client.MergeFrom(stored)); err != nil {
+			return false, fmt.Errorf("patching nodeclaim %q, %w", nc.Name, err)
+		}
+	}
+
+	// If the reservation expired before the NodeClaim became registered, there may not be a Node on the cluster. Note
+	// that there should never be duplicate Nodes for a given NodeClaim, but handling this user-induced error is more
+	// straightforward than handling the duplicate error.
+	nodes, err := nodeclaimutils.AllNodesForNodeClaim(ctx, c.kubeClient, nc)
```

Comment:
If we get a NotFound error, I think you're saying that this should be a requeue because there's a tiny race that could occur that would cause us to update the label after the NodeClaim registration code had already propagated the reserved label to the node -- should we surface that as an error, though? Or is it best to just treat this as a chance to requeue? AFAIK, this should just get updated and handled later, right?

Reply:
This comment was moreso meant to indicate why I'm not requeing for either of these errors here. For a missing node, we'll get to it on the next iterations, and for duplicate nodes, it doesn't make sense not to handle it.

Reply:
I guess my callout was more that you are going to print the error every time -- is this something you want to do?

Reply:
This function never returns `NotFound`, if there are no Nodes for the NodeClaim's provider ID we just get an empty list.